tests/test_array.h

```diff
@@ -0,0 +1,179 @@
+/*************************************************************************/
+/* Copyright (c) 2007-2020 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2020 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef TEST_ARRAY_H
+#define TEST_ARRAY_H
+
+#include "core/object/class_db.h"
+#include "core/object/script_language.h"
+#include "core/templates/hashfuncs.h"
+#include "core/templates/vector.h"
+#include "core/variant/array.h"
+#include "core/variant/container_type_validate.h"
+#include "core/variant/variant.h"
+#include "tests/test_macros.h"
+
+namespace TestArray {
+
+TEST_CASE("[Array] Size, Clear, and Empty") {
+	Array arr;
+	CHECK(arr.size() == 0);
+	CHECK(arr.empty());
+	arr.push_back(1);
+	CHECK(arr.size() == 1);
+	arr.clear();
+	CHECK(arr.empty());
+	CHECK(arr.size() == 0);
+}
+
+TEST_CASE("[Array] Assignment and Compare Operators") {
+	Array arr1;
+	Array arr2;
+	arr1.push_back(1);
+	CHECK(arr1 != arr2);
+	CHECK(arr1 > arr2);
+	CHECK(arr1 >= arr2);
+	arr2.push_back(2);
+	CHECK(arr1 != arr2);
+	CHECK(arr1 < arr2);
+	CHECK(arr1 <= arr2);
+	CHECK(arr2 > arr1);
+	CHECK(arr2 >= arr1);
+	Array arr3 = arr2;
+	CHECK(arr3 == arr2);
+}
+
+TEST_CASE("[Array] append_array") {
+	Array arr1;
+	Array arr2;
+	arr1.push_back(1);
+	arr1.append_array(arr2);
+	CHECK(arr1.size() == 1);
+	arr2.push_back(2);
+	arr1.append_array(arr2);
+	CHECK(arr1.size() == 2);
+	CHECK(int(arr1[0]) == 1);
+	CHECK(int(arr1[1]) == 2);
```

Comment:
This should work, but not 100% sure.

```suggestion
	CHECK(arr1[0] == 1);
	CHECK(arr1[1] == 2);
```

The same applies to other assertions.

Reply:
This does not work. arr1[0] returns the Variant object and so the assertion will give an error because you are trying to compare a variant and integer.

Reply:
Yes, as it turns out, `Variant` doesn't support comparison against primitive data types. This usually works for other `Variant` compatible types such as `Array`: `Variant() == Array()` would evaluate to `false`, and even works for string literals.

I think this is something which can be implemented in `Variant` directly, but this is certainly outside the scope of this PR.